### DISCUSSION TITLE
Any tips for debugging connection pool timeouts?

### DISCUSSION THREAD

**Discussion by RobertCraigie** [encode/httpx#2940] [Q&A]
Hi, I'm a maintainer of the [`openai-python`](https://github.com/openai/openai-python) package which recently switched over to using httpx internally 🙏

We're seeing some reports from users regarding connection pool timeouts: https://github.com/openai/openai-python/issues/769. Unfortunately no users have been able to share a reproduction yet and I haven't managed to reproduce the problem myself.

From looking through debug logs for `httpx` and `httpcore` I haven't seen anything explicitly mentioning the connection pool handling, just when connections are started / closed. I've also gone through our code and didn't find any obvious place where we're leaving connections open.

Any tips for debugging this would be greatly appreciated!

**Comment by karpetrosyan** [comment]
Hi @RobertCraigie!

Here are some suggestions:

- Update to the latest httpx and httpcore versions
- Do not handle "asyncio.Cancelled" exceptions (Sometimes it can break the client's state).
- Use https://github.com/encode/httpcore/pull/702 branch for connection pool monitoring.

If there is a timeout issue, it comes from `httpcore`, so you can see the existing `httpcore` issues and maybe you can find something interesting.

**Reply by RobertCraigie** [reply]
Thanks Kar, any idea when that PR would be merged / released?

**Reply by karpetrosyan** [reply]
I can't really say when it's going to be released, but I think it can't help you without the minimum code that can reproduce the issue.

**Reply by RobertCraigie** [reply]
Okay thanks, your branch has been very helpful in debugging :)

I still haven't managed to reproduce the reported connection pool timeouts but I am seeing some strange behaviour with read timeouts randomly happening - https://github.com/encode/httpx/discussions/2941
